File: 0735.行星碰撞/0735-行星碰撞.rs
impl Solution {
    pub fn asteroid_collision(nums: Vec<i32>) -> Vec<i32> {
        let mut stack = vec![];
        for num in nums {
            if stack.is_empty() || stack[stack.len()-1] < 0 || num > 0 {
                stack.push(num);
            } else if num < 0 {
                let mut push_num = true;
                while !stack.is_empty() && push_num {
                    let last = stack.len() - 1;
                    if stack[last] > 0 {
                        if stack[last] > num.abs() {
                            push_num = false;
                        } else if stack[last] == num.abs() {
                            stack.pop();
                            push_num = false;
                        } else if stack[last] < num.abs() {
                            stack.pop();
                        }
                    } else if stack[last] < 0 {
                        break;
                    }
                }
                if push_num {
                    stack.push(num);
                }
            }
        }

        stack
    }
}

File: 0735.行星碰撞/0735-行星碰撞.py
class Solution:
    def asteroidCollision(self, nums: List[int]) -> List[int]:
        stack = []
        for n in nums:
            if not stack or n > 0:
                stack.append(n)
            else:
                keep_n = True
                while stack and keep_n:
                    if stack[-1] > 0:
                        if stack[-1] < abs(n):
                            stack.pop()
                        elif stack[-1] == abs(n):
                            stack.pop()
                            keep_n = False
                        else:
                            keep_n = False
                    else:
                        break
                if keep_n:
                    stack.append(n)
        return stack
